lib/ukargparse: Use mutable strings in tests

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]


### Additional configuration

None.

### Description of changes

Tests in libukarparse pass `ukargnparse()` with string literals. As string literals are stored in `.rodata`, and `ukargnparse()` performs in-place modification of parsed strings, this causes a data abort in KVM / Arm64. The only reason it works on x86 is because segments attributes are not properly set and `.text` and `.rodata` are not read-only (#414) 😛 This commit adds a macro to convert string literals into string buffers.

michpappas committed:
lib/ukargparse: Use mutable strings in tests

As uk_argnparse() modifies input strings in-place, it is required that
argb points to a mutable string. Replace string literals with mutable
arrays of chars.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

craciunoiuc reviewed (approved):
@michpappas @razvand 

All good here. To be honest, that statement expression[1] in the macro left me baffled. I didn't even know how to look it up. No need to worry though, I looked it up and it seems that it works on `clang` and the `intel` compiler now too.

[1] https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@gmail.com>

razvand commented:
Hi, @michpappas . I tested your proposed update and test fail:

```
test: ukargparse->parse_extra_whitespaces
    :   expected `argc` to be 3 but was 1 ............................................... [ FAILED ]
    :     in test_ukargparse_parse.c]:93]
test: ukargparse->parse_null
    :   expected `argc` to be 0 and was 0 ............................................... [ PASSED ]
    :   expected `arg_out` to be 0 and was 0 ............................................ [ PASSED ]
test: ukargparse->parse_quotes
    :   expected `argc` to be 5 but was 1 ............................................... [ FAILED ]
    :     in test_ukargparse_parse.c]:112]
test: ukargparse->parse_space_separated
    :   expected `argc` to be 4 but was 1 ............................................... [ FAILED ]
    :     in test_ukargparse_parse.c]:63]
```

This is because the `argb` argument to `uk_argnparse()` is not properly initialized. I added some debug printing in `lib/ukargparse/argparse.c` and got this:

```
test: ukargparse->parse_extra_whitespaces
[    0.137790] dbg:  [libukargparse] <argparse.c @  100> character is 0x90
[    0.139465] dbg:  [libukargparse] <argparse.c @  101> argc is 1
[    0.140919] dbg:  [libukargparse] <argparse.c @  100> character is 0xff
[    0.142523] dbg:  [libukargparse] <argparse.c @  101> argc is 1
[    0.143959] dbg:  [libukargparse] <argparse.c @  100> character is 0xf1
[    0.145463] dbg:  [libukargparse] <argparse.c @  101> argc is 1
[    0.146945] dbg:  [libukargparse] <argparse.c @  100> character is 0x03
[    0.148366] dbg:  [libukargparse] <argparse.c @  101> argc is 1
    :   expected `argc` to be 3 but was 1 ............................................... [ FAILED ]
    :     in test_ukargparse_parse.c]:93]
test: ukargparse->parse_null
    :   expected `argc` to be 0 and was 0 ............................................... [ PASSED ]
    :   expected `arg_out` to be 0 and was 0 ............................................ [ PASSED ]
test: ukargparse->parse_quotes
[    0.159128] dbg:  [libukargparse] <argparse.c @  100> character is 0x90
[    0.160523] dbg:  [libukargparse] <argparse.c @  101> argc is 1
[    0.161801] dbg:  [libukargparse] <argparse.c @  100> character is 0xff
[    0.164244] dbg:  [libukargparse] <argparse.c @  101> argc is 1
[    0.166168] dbg:  [libukargparse] <argparse.c @  100> character is 0xf1
[    0.167751] dbg:  [libukargparse] <argparse.c @  101> argc is 1
[    0.169040] dbg:  [libukargparse] <argparse.c @  100> character is 0x03
[    0.170589] dbg:  [libukargparse] <argparse.c @  101> argc is 1
    :   expected `argc` to be 5 but was 1 ............................................... [ FAILED ]
```

So the construct you used doesn't seem to work. I did some assembly debugging but didn't figure out what is does and why that is the case.

However, given that it's pretty complicated (and weird) I suggest we simply used array strings for the tests, such as in the patch below:

```
diff --git a/lib/ukargparse/tests/test_ukargparse_parse.c b/lib/ukargparse/tests/test_ukargparse_parse.c
index 662ffbc3..9be6b9e5 100644
--- a/lib/ukargparse/tests/test_ukargparse_parse.c
+++ b/lib/ukargparse/tests/test_ukargparse_parse.c
@@ -40,19 +40,9 @@
 
 #include <uk/argparse.h>
 
-/* Converts a string literal into mutable string buffer. */
-#define STR_BUF(str) ({					\
-	char buf[64];					\
-	UK_ASSERT(strlen(str) < ARRAY_SIZE(buf));	\
-	strcpy(buf, str);				\
-	buf;						\
-})
-
 UK_TESTCASE(ukargparse, parse_space_separated)
 {
-	char *arg_str = {
-		STR_BUF("some/path/file -test 1 --test-this")
-	};
+	char arg_str[] = "some/path/file -test 1 --test-this";
 	static const char * const arg_ex[] = {"some/path/file",
 		"-test", "1", "--test-this"};
 
@@ -81,9 +71,7 @@ UK_TESTCASE(ukargparse, parse_null)
 
 UK_TESTCASE(ukargparse, parse_extra_whitespaces)
 {
-	char *arg_str = {
-		STR_BUF("some\t\t\n\t\tseparated\n\t\t  \n  string")
-	};
+	char arg_str[] = "some\t\t\n\t\tseparated\n\t\t  \n  string";
 	static const char * const arg_ex[] = {"some", "separated", "string"};
 
 	// null-terminated array of parsed arguments
@@ -100,9 +88,7 @@ UK_TESTCASE(ukargparse, parse_extra_whitespaces)
 
 UK_TESTCASE(ukargparse, parse_quotes)
 {
-	char *arg_str = {
-		STR_BUF("\"\targ_a\" ' arg_b' \"arg'c\" arg_d'' arg\"_\"e\"'\"")
-	};
+	char arg_str[] = "\"\targ_a\" ' arg_b' \"arg'c\" arg_d'' arg\"_\"e\"'\"";
 	static const char * const arg_ex[] = {"\targ_a", " arg_b", "arg'c", "arg_d", "arg_e'"};
 
 	// null-terminated array of parsed arguments
 ```

craciunoiuc commented:
Hmm, strange, I swear it worked last time I tried it

gabrielmocanu commented:
@michpappas  any updates on this?

michpappas commented:
Hi @razvand. I couldn't reproduce the problem, but in any case the approach you propose is way more elegant, so thanks for that! 🙂 I updated the PR accordingly, and rebased from staging. If you are planning to test, don't forget to cherry-pick the fixes in uktest from #602 and #606.

craciunoiuc reviewed (approved):
Now it looks double good 😄 

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@gmail.com>

razvand reviewed (approved):
Thanks, @michpappas 

Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
